### PR TITLE
On-demand database exports

### PR DESCRIPTION
**Pull request by kavorite**
The intent is that this endpoint's eligibility as upstream functionality is mutually exclusive with #71. One or the other should be decided upon for implementation. I'm in favor of this model personally ( it is totally not because I got to use previously unexplored coroutine hacks ).

### CODE REVIEW

**Comment by kavorite** [comment]
This branch lacks compression, but it probably wouldn't be that difficult to implement

**Comment by kavorite** [comment]
update: the hacks were too hacky

File: masterbase/app.py [hunk 1]
```diff
@@ -165,6 +166,20 @@ async def demodata(request: Request, api_key: str, session_id: str) -> Redirect:
...
+    filename = f"demo_sessions-{datetime.now()}.csv"
+    return Stream(
+        db_export_cached(engine, data.table, data.max_age),
+        headers={
+            "Content-Type": "text/csv",
+            "Content-Disposition": f"attachment; filename={filename}",
+        },
+    )
+
+
```
Comment (jayceslesar) [inline masterbase/app.py:182]: would be nice to add a test for this given we have the infra

Comment (kavorite) [reply masterbase/app.py:182]: how mock demo_sessions data

Comment (kavorite) [reply masterbase/app.py]: i didn't bother, used the reports table instead

**Comment by kavorite** [comment]
Implementing the integration test left only one more housekeeping item: finding a way to prevent threads reading and writing cached results from racing. That includes unnecessarily overwriting the cache file, reading to the end of an incomplete cache file as it is still being written, and a panoply of other potential race conditions. 

There are ways of mitigating this. For instance, moving a file on POSIX systems is atomic, which can be leveraged to prevent contention. Alternatively, rendezvous with the database could be synchronized to ensure no requester can export while another is running.

Still, concurrent programming is complicated. Even if all of those synchronization issues were easy to resolve (they aren't), those solutions involve multiple threads doing random I/O, which needs to be prevented to preserve throughput. Every solution I could propose requires simplicity or ease of use compromises. 

I'm declaring this a case of premature optimization. The most straightforward approach, by far, is not to hand out analyst authorization to anyone found to abuse this functionality. If we encounter any issues, we can implement an endpoint to allow requesters to check how many exports are in progress, guard them with a semaphore, or accept a list of columns to export.

Until such measures are necessary, this branch begins a new export job every time a snapshot is requested. It delegates synchronization to Postgres, simplifying its implementation and preserving its correctness by privatizing results. It should still be reasonably performant, provided that people don't start too many simultaneous `COPY` routines.

File: tests/test_api.py [hunk 2]
```diff
@@ -42,7 +44,9 @@ def test_client(steam_id: str, api_key: str) -> Iterator[TestClient[Litestar]]:
         yield client
 
         with app.state.engine.connect() as conn:
-            sql = "DELETE FROM demo_sessions;"
+            sql = "TRUNCATE TABLE reports CASCADE;"
+            conn.execute(sa.text(sql))
+            sql = "TRUNCATE TABLE demo_sessions CASCADE;"
             conn.execute(sa.text(sql))
             sql = "DELETE FROM api_keys WHERE api_key = :api_key;"
             conn.execute(sa.text(sql), {"api_key": api_key})
```
Comment (jayceslesar) [inline tests/test_api.py:52]: awesome change, this is super clean

**Pull request merged**